File: src/parser.rs
//! Parse SVCMS-formatted commit messages

use regex::Regex;
use anyhow::Result;
use crate::SvcmsCommit;

lazy_static::lazy_static! {
    // Header pattern: <type>(<scope>): <summary>
    static ref HEADER_PATTERN: Regex = Regex::new(
        r"^(\w+)(?:\(([^)]+)\))?:\s*(.+)"
    ).unwrap();
    
    // Footer patterns for extracting metadata
    static ref CONTEXT_PATTERN: Regex = Regex::new(
        r"(?m)^Context:\s*(.+)$"
    ).unwrap();
    
    static ref REFS_PATTERN: Regex = Regex::new(
        r"(?m)^Refs?:\s*(.+)$"
    ).unwrap();
    
    static ref MEMORY_PATTERN: Regex = Regex::new(
        r"(?m)^Memory:\s*(.+)$"
    ).unwrap();
    
    static ref LOCATION_PATTERN: Regex = Regex::new(
        r"(?m)^Location:\s*(.+)$"
    ).unwrap();
    
    static ref TAGS_PATTERN: Regex = Regex::new(
        r"(?m)^Tags?:\s*(.+)$"
    ).unwrap();
}

/// Parse a commit message following SVCMS format
pub fn parse_commit_message(
    sha: &str, 
    message: &str, 
    timestamp: chrono::DateTime<chrono::Utc>
) -> Result<Option<SvcmsCommit>> {
    let lines: Vec<&str> = message.lines().collect();
    if lines.is_empty() {
        return Ok(None);
    }
    
    // Parse header
    let header = lines[0];
    let captures = match HEADER_PATTERN.captures(header) {
        Some(c) => c,
        None => return Ok(None), // Not an SVCMS commit
    };
    
    let commit_type = captures.get(1)
        .map(|m| m.as_str().to_string())
        .unwrap_or_default();
    
    // Check if this is an SVCMS-extended type
    if !is_valid_svcms_type(&commit_type) {
        return Ok(None);
    }
    
    let scope = captures.get(2).map(|m| m.as_str().to_string());
    let summary = captures.get(3)
        .map(|m| m.as_str().to_string())
        .unwrap_or_default();
    
    // Extract body (everything between header and footers)
    let body = extract_body(&lines);
    
    // Extract footers
    let full_message = lines.join("\n");
    let context = extract_field(&full_message, &CONTEXT_PATTERN);
    let refs = extract_refs(&full_message);
    let memory = extract_field(&full_message, &MEMORY_PATTERN);
    let location = extract_field(&full_message, &LOCATION_PATTERN);
    let tags = extract_tags(&full_message);
    
    Ok(Some(SvcmsCommit {
        sha: sha.to_string(),
        commit_type,
        scope,
        summary,
        body,
        memory,
        location,
        context,
        refs,
        tags,
        timestamp,
    }))
}

/// Check if the commit type is valid according to SVCMS
fn is_valid_svcms_type(commit_type: &str) -> bool {
    matches!(
        commit_type,
        // Standard Conventional Commits
        "feat" | "fix" | "fixed" | "docs" | "style" | "refactor" | 
        "perf" | "test" | "build" | "ci" | "chore" |
        // SVCMS Knowledge Types
        "learned" | "insight" | "context" | "decision" | "decided" | "memory" |
        // SVCMS Collaboration Types
        "discussed" | "explored" | "attempted" |
        // SVCMS Meta Types
        "workflow" | "preference" | "pattern"
    )
}

/// Extract the body content (between header and footers)
fn extract_body(lines: &[&str]) -> Option<String> {
    if lines.len() <= 1 {
        return None;
    }
    
    let mut body_lines = Vec::new();
    let mut in_body = false;
    
    for line in lines.iter().skip(1) {
        // Empty line after header starts the body
        if !in_body && line.trim().is_empty() {
            in_body = true;
            continue;
        }
        
        // Check if we've hit a footer
        if in_body && is_footer_line(line) {
            break;
        }
        
        if in_body {
            body_lines.push(*line);
        }
    }
    
    if body_lines.is_empty() {
        None
    } else {
        Some(body_lines.join("\n").trim().to_string())
    }
}

/// Check if a line is a footer line
fn is_footer_line(line: &str) -> bool {
    line.starts_with("Context:") ||
    line.starts_with("Refs:") ||
    line.starts_with("Ref:") ||
    line.starts_with("Memory:") ||
    line.starts_with("Location:") ||
    line.starts_with("Tags:") ||
    line.starts_with("Tag:")
}

/// Extract a single field value from the message
fn extract_field(message: &str, pattern: &Regex) -> Option<String> {
    pattern.captures(message)
        .and_then(|cap| cap.get(1))
        .map(|m| m.as_str().trim().to_string())
}

/// Extract references as a vector
fn extract_refs(message: &str) -> Vec<String> {
    extract_field(message, &REFS_PATTERN)
        .map(|refs_str| {
            refs_str.split(',')
                .map(|r| r.trim().to_string())
                .filter(|r| !r.is_empty())
                .collect()
        })
        .unwrap_or_default()
}

/// Extract tags as a vector
fn extract_tags(message: &str) -> Vec<String> {
    extract_field(message, &TAGS_PATTERN)
        .map(|tags_str| {
            tags_str.split(',')
                .map(|t| t.trim().to_string())
                .filter(|t| !t.is_empty())
                .collect()
        })
        .unwrap_or_default()
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::Utc;
    
    #[test]
    fn test_parse_standard_commit() {
        let message = "feat(auth): add JWT authentication";
        let result = parse_commit_message("abc123", message, Utc::now()).unwrap();
        
        assert!(result.is_some());
        let commit = result.unwrap();
        assert_eq!(commit.commit_type, "feat");
        assert_eq!(commit.scope, Some("auth".to_string()));
        assert_eq!(commit.summary, "add JWT authentication");
    }
    
    #[test]
    fn test_parse_svcms_commit_with_memory() {
        let message = r#"learned(api): rate limiting resets at minute boundaries

Discovered through testing that the API rate limiter uses fixed minute
boundaries rather than a rolling 60-second window.

Context: Staff Scheduling API integration
Refs: #87, src/api/client.ts
Memory: API rate limit resets at :00 seconds of each minute
Location: src/api/CLAUDE.md
Tags: api, rate-limiting, retry-strategy"#;
        
        let result = parse_commit_message("def456", message, Utc::now()).unwrap();
        
        assert!(result.is_some());
        let commit = result.unwrap();
        assert_eq!(commit.commit_type, "learned");
        assert_eq!(commit.scope, Some("api".to_string()));
        assert_eq!(commit.summary, "rate limiting resets at minute boundaries");
        assert!(commit.body.is_some());
        assert_eq!(commit.context, Some("Staff Scheduling API integration".to_string()));
        assert_eq!(commit.memory, Some("API rate limit resets at :00 seconds of each minute".to_string()));
        assert_eq!(commit.location, Some("src/api/CLAUDE.md".to_string()));
        assert_eq!(commit.refs.len(), 2);
        assert_eq!(commit.tags.len(), 3);
    }
    
    #[test]
    fn test_parse_non_svcms_commit() {
        let message = "random commit message without proper format";
        let result = parse_commit_message("ghi789", message, Utc::now()).unwrap();
        assert!(result.is_none());
    }
    
    #[test]
    fn test_parse_empty_commit() {
        let message = r#"decided(architecture): use event-driven pattern

Context: Design discussion with Corrado
Memory: All state changes through events
Tags: architecture, events"#;
        
        let result = parse_commit_message("jkl012", message, Utc::now()).unwrap();
        
        assert!(result.is_some());
        let commit = result.unwrap();
        assert_eq!(commit.commit_type, "decided");
        assert_eq!(commit.scope, Some("architecture".to_string()));
        assert!(commit.body.is_none()); // No body, just footers
        assert_eq!(commit.memory, Some("All state changes through events".to_string()));
    }
}

File: mcp-server/src/svcms/parser.ts
/**
 * SVCMS Commit Message Parser
 * 
 * Parses commit messages following the SVCMS specification.
 * Ported from Rust implementation with TypeScript enhancements.
 */

import { logger } from '../utils/logger.js';
import type { ParsedSvcmsCommit } from '../types.js';

export class SvcmsParser {
  // SVCMS commit types (from specification)
  private readonly SVCMS_TYPES = new Set([
    // Standard types
    'feat', 'fix', 'docs', 'style', 'refactor', 'perf', 'test', 'build', 'ci', 'chore',
    // Knowledge types  
    'learned', 'insight', 'context', 'decision', 'memory',
    // Collaboration types
    'discussed', 'explored', 'attempted',
    // Meta types
    'workflow', 'preference', 'pattern',
    // Legacy variants
    'decided' // alias for decision
  ]);

  // Two-tier format regex: category.type(scope): summary
  private readonly TWO_TIER_REGEX = /^([a-z]+)\.([a-z]+)(?:\(([^)]+)\))?\s*:\s*(.+)$/;
  
  // Legacy format regex: type(scope): summary  
  private readonly LEGACY_REGEX = /^([a-z]+)(?:\(([^)]+)\))?\s*:\s*(.+)$/;

  /**
   * Parse SVCMS commit message
   */
  parse(message: string): ParsedSvcmsCommit | undefined {
    try {
      const lines = message.split('\n');
      const firstLine = lines[0]?.trim();
      
      if (!firstLine) {
        logger.debug('Empty commit message');
        return undefined;
      }

      // Try two-tier format first
      let match = firstLine.match(this.TWO_TIER_REGEX);
      if (match) {
        const [, category, type, scope, summary] = match;
        
        if (!this.SVCMS_TYPES.has(type)) {
          logger.debug(`Unknown SVCMS type: ${type}`);
          return undefined;
        }

        return this.parseCommitBody(lines.slice(1), {
          category,
          type,
          scope,
          summary: summary.trim()
        });
      }

      // Try legacy format
      match = firstLine.match(this.LEGACY_REGEX);
      if (match) {
        const [, type, scope, summary] = match;
        
        if (!this.SVCMS_TYPES.has(type)) {
          logger.debug(`Unknown SVCMS type: ${type}`);
          return undefined;
        }

        return this.parseCommitBody(lines.slice(1), {
          type,
          scope,
          summary: summary.trim()
        });
      }

      logger.debug('Commit message does not match SVCMS format');
      return undefined;

    } catch (error) {
      logger.warn('Failed to parse commit message:', error);
      return undefined;
    }
  }

  /**
   * Parse commit body for SVCMS metadata
   */
  private parseCommitBody(
    bodyLines: string[], 
    header: Partial<ParsedSvcmsCommit>
  ): ParsedSvcmsCommit {
    const result: ParsedSvcmsCommit = {
      ...header
    } as ParsedSvcmsCommit;

    let bodyContent = '';
    
    for (const line of bodyLines) {
      const trimmed = line.trim();
      
      // Skip empty lines at the start
      if (!trimmed && !bodyContent) {
        continue;
      }

      // Check for metadata fields
      if (trimmed.startsWith('Context:')) {
        result.context = trimmed.substring(8).trim();
        continue;
      }

      if (trimmed.startsWith('Refs:')) {
        result.refs = this.parseRefs(trimmed.substring(5).trim());
        continue;
      }

      if (trimmed.startsWith('Memory:')) {
        result.memory = trimmed.substring(7).trim();
        continue;
      }

      if (trimmed.startsWith('Location:')) {
        result.location = trimmed.substring(9).trim();
        continue;
      }

      if (trimmed.startsWith('Tags:')) {
        result.tags = this.parseTags(trimmed.substring(5).trim());
        continue;
      }

      // Skip Claude Code attribution lines
      if (trimmed.includes('Generated with [Claude Code]') || 
          trimmed.includes('Co-Authored-By: Claude') ||
          trimmed.startsWith('🤖') ||
          trimmed.startsWith('🚀')) {
        continue;
      }

      // Accumulate body content
      if (bodyContent || trimmed) {
        bodyContent += (bodyContent ? '\n' : '') + line;
      }
    }

    if (bodyContent.trim()) {
      result.body = bodyContent.trim();
    }

    logger.debug('Parsed SVCMS commit:', result);
    return result;
  }

  /**
   * Parse references field (issues, docs, commits)
   */
  private parseRefs(refsString: string): string[] {
    if (!refsString) return [];
    
    return refsString
      .split(',')
      .map(ref => ref.trim())
      .filter(ref => ref.length > 0);
  }

  /**
   * Parse tags field
   */
  private parseTags(tagsString: string): string[] {
    if (!tagsString) return [];
    
    return tagsString
      .split(',')
      .map(tag => tag.trim())
      .filter(tag => tag.length > 0);
  }

  /**
   * Validate SVCMS commit structure
   */
  validateCommit(parsed: ParsedSvcmsCommit): {
    valid: boolean;
    errors: string[];
    warnings: string[];
  } {
    const errors: string[] = [];
    const warnings: string[] = [];

    // Required fields
    if (!parsed.type) {
      errors.push('Missing commit type');
    }

    if (!parsed.summary) {
      errors.push('Missing commit summary');
    }

    // Summary length check
    if (parsed.summary && parsed.summary.length > 72) {
      warnings.push('Summary exceeds 72 characters');
    }

    // Memory field for knowledge commits
    if (['learned', 'insight', 'decision', 'memory'].includes(parsed.type) && !parsed.memory) {
      warnings.push('Knowledge commits should include Memory field');
    }

    // Location field validation
    if (parsed.location && !this.isValidLocation(parsed.location)) {
      warnings.push('Location field should be a valid file path');
    }

    return {
      valid: errors.length === 0,
      errors,
      warnings
    };
  }

  /**
   * Check if location is a valid file path
   */
  private isValidLocation(location: string): boolean {
    // Basic validation for CLAUDE.md paths
    return location.includes('CLAUDE.md') || 
           location.startsWith('./') || 
           location.startsWith('~/') ||
           location.startsWith('src/');
  }

  /**
   * Generate commit suggestion based on diff analysis
   */
  generateCommitSuggestion(diff: string, context?: {
    scope?: string;
    recentCommits?: ParsedSvcmsCommit[];
  }): {
    category: string;
    type: string;
    scope?: string;
    summary: string;
    confidence: number;
  } {
    // Simple heuristics for commit suggestion
    const diffLower = diff.toLowerCase();
    
    // Detect type based on diff content
    let type = 'chore';
    let category = 'standard';
    let confidence = 0.5;

    if (diffLower.includes('test') || diffLower.includes('spec')) {
      type = 'test';
      confidence = 0.8;
    } else if (diffLower.includes('fix') || diffLower.includes('bug')) {
      type = 'fix';
      confidence = 0.9;
    } else if (diffLower.includes('feat') || diffLower.includes('add')) {
      type = 'feat';
      confidence = 0.8;
    } else if (diffLower.includes('refactor')) {
      type = 'refactor';
      confidence = 0.9;
    } else if (diffLower.includes('perf')) {
      type = 'perf';
      confidence = 0.8;
    } else if (diffLower.includes('doc')) {
      type = 'docs';
      confidence = 0.8;
    }

    // Detect knowledge commits
    if (diffLower.includes('todo') || 
        diffLower.includes('fixme') ||
        diffLower.includes('memory:')) {
      category = 'knowledge';
      type = 'learned';
      confidence = 0.7;
    }

    // Generate summary based on changes
    const summary = this.generateSummaryFromDiff(diff);

    return {
      category,
      type,
      scope: context?.scope,
      summary,
      confidence
    };
  }

  /**
   * Generate summary from diff content
   */
  private generateSummaryFromDiff(diff: string): string {
    const lines = diff.split('\n');
    const additions = lines.filter(line => line.startsWith('+')).length;
    const deletions = lines.filter(line => line.startsWith('-')).length;
    
    if (additions > deletions * 2) {
      return 'add new functionality';
    } else if (deletions > additions * 2) {
      return 'remove unused code';
    } else {
      return 'update implementation';
    }
  }
}